### PR TITLE
[feat] Add Middlewares

### PR DESCRIPTION
**Pull request by warittornc**
## Description
- Added middleware for caching results by signature and request ID and external ID
- Added RedisCache
- Minor refactor and cleanup

### CODE REVIEW

File: app/utils/log_config.py [hunk 1]
```diff
@@ -27,3 +27,6 @@ def init_loggers(logger_name: str = "pds_gateway_log", log_level: str = "DEBUG")
 
     # Adds the console handler to the logger
     logger.addHandler(ch)
+
+    log = logging.getLogger("pds_gateway_log")
+    return log
```
Comment (taobun) [inline app/utils/log_config.py:32]: add return type on function?

File: app/main.py [hunk 2]
```diff
@@ -1,108 +1,163 @@
...
+            report.response_code = e.status_code
+            report.error_type = e.error_type
+            report.error_msg = str(e)
+        except Exception as e:
+            report.response_code = 500
+            report.error_type = VerifyErrorType.UNKNOWN.value
+            report.error_msg = str(e)
+        finally:
+            # Save report to db if db is enabled
+            if db:
```
Comment (taobun) [inline app/main.py:79]: should be `verify_db`?

File: app/main.py [hunk 3]
```diff
@@ -1,108 +1,163 @@
...
+async def request_data(request: Request, _: None = Depends(verify_request)) -> Any:
+    """Requests data from the premium data source"""
+    report = ProviderResponseReport(response_code=200)
     try:
-        output = await app.state.adapter.unified_call(dict(request.query_params))
-        if settings.MODE == "production":
-            # Cache data
-            app.state.cache_data.set_data(get_band_signature_hash(request.headers), output)
-        return output
+        return await adapter.unified_call(dict(request.query_params))
```
Comment (taobun) [inline app/main.py:88]: Should we check result of verify before call adapter?

File: app/main.py [hunk 4]
```diff
@@ -1,108 +1,163 @@
...
 async def get_status_report() -> StatusReport:
-    """Gets a status report: gateway info, latest request and latest failed request"""
+    """Gets a status report that contains the latest reports"""
     gateway_info = GatewayInfo(
         allow_data_source_ids=settings.ALLOWED_DATA_SOURCE_IDS,
         max_delay_verification=settings.MAX_DELAY_VERIFICATION,
     )
 
-    if app.state.db:
+    if db:
```
Comment (taobun) [inline app/main.py:116]: which db?

File: app/main.py [hunk 5]
```diff
@@ -1,108 +1,163 @@
...
+
+@report_app.get("/latest_failed")
+async def get_failed_status_report() -> StatusReport:
+    """Gets a status report that contains the latest failed reports"""
+    gateway_info = GatewayInfo(
+        allow_data_source_ids=settings.ALLOWED_DATA_SOURCE_IDS,
+        max_delay_verification=settings.MAX_DELAY_VERIFICATION,
     )
+
+    if db:
```
Comment (taobun) [inline app/main.py:143]: ditto

File: app/main.py [hunk 6]
```diff
@@ -1,108 +1,163 @@
...
+        report.response_code = 500
+        report.error_msg = str(e)
         raise HTTPException(
             status_code=500,
-            detail={"error_msg": f"{e}"},
+            detail=str(e),
         )
+    finally:
+        if provider_response_db:
+            provider_response_db.save(report)
```
Comment (taobun) [inline app/main.py:105]: finally work on re raise?

**Review by taobun — CHANGES_REQUESTED** [review]

File: app/main.py (hunk 6 above)
Comment (taobun) [reply app/main.py:105]: worked just read doc and it works

File: app/main.py [hunk 7]
```diff
@@ -1,108 +1,165 @@
...
-from app.utils.cache import Cache
-from app.utils.helper import is_data_source_id_allowed, verify_request_from_bandchain, get_band_signature_hash
+from app.report.models import StatusReport, GatewayInfo, VerifyReport, ProviderResponseReport, RequestReport
+from app.utils.cache import LocalCache, RedisCache
+from app.utils.helper import is_data_source_id_allowed, verify_request_from_bandchain
 from app.utils.log_config import init_loggers
 from app.utils.types import VerifyErrorType
 
+# Setup apps
 app = FastAPI()
```
Comment (RogerKSI) [inline app/main.py:21]: do we still use this app?

File: app/report/models.py [hunk 8]
```diff
@@ -50,5 +53,6 @@ class GatewayInfo(BaseModel):
 
 class StatusReport(BaseModel):
```
Comment (RogerKSI) [inline app/report/models.py:54]: Not sure what's the purpose of `Report`? seems there is no common.
also, this one should inherit from `Report` as well?

File: app/main.py (hunk 7 above)
Comment (warittornc) [reply app/main.py:21]: yes, the other apps are mounted to this app

File: app/report/models.py (hunk 8 above)
Comment (warittornc) [reply app/report/models.py:54]: `Report` is used for type hinting to give reports a common parent. `StatusReport` is an aggregation of other reports  + gateway info. Maybe a name change might be better for clarity?

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]
LGTM let's try on testnet data source

**Review by RogerKSI — APPROVED** [review]

**Pull request merged**